crates/ty_python_semantic/src/semantic_index/reachability_constraints.rs

```diff
@@ -678,6 +678,35 @@ impl ReachabilityConstraints {
                 let ty = infer_expression_type(db, test_expr);
                 ty.bool(db).negate_if(!predicate.is_positive)
             }
+            PredicateNode::ReturnsNever(test_expr) => {
+                let ty = infer_expression_type(db, test_expr);
+                if let Type::FunctionLiteral(function_literal) = ty {
+                    let returns_never =
+                        if function_literal
+                            .signature(db)
+                            .overloads
+                            .iter()
+                            .all(|overload| {
+                                // HACK: for now, require that *all* overloads are annotated with
+                                // returning `Never`
+                                // Ideally, if only some overloads return `Never`, we should consider
+                                // the types of the arguments.
+                                overload.return_ty.is_some_and(|return_type| {
+                                    return_type.is_equivalent_to(db, Type::Never)
+                                })
+                            })
+                        {
+                            Truthiness::AlwaysTrue
+                        } else {
+                            Truthiness::AlwaysFalse
+                        };
+                    returns_never.negate_if(!predicate.is_positive)
+                } else {
+                    // Should I add a panic here?
```

Comment:
No we shouldn't add a panic here, we should silently return something and let type checking emit a diagnostic for a bad call.

It's an interesting question what we should return. In principle it makes sense to return `AlwaysTrue` (calling a non-callable will raise an error and thus terminate control flow). But we don't do that for other diagnostics that are high-confidence will-raise-an-error, so it seems inconsistent to do it here. Thus I think we should probably return `AlwaysFalse`.

Reply:
That makes sense. I changed this to return `AlwaysFalse`. It seems earlier I had the impression that this branch should be unreachable, hence the question about adding a panic. But I now get that that impression is clearly wrong, since semantic indexing will happily add a constraint here for `random_thing_which_is_not_a_function()`, by looking at just the AST :)